### PR TITLE
Add a "Location" to support S3 and local paths

### PR DESCRIPTION
**Pull request by jder**
This lets us:
* Add structured info for locations (eg s3 endpoint URLs)
* Still use raw strings most of the time
* Get some type safety to know if we've correctly resolved these against the root before opening.
* (In a future PR) detect the need for "spawn" multiprocessing (required for S3 but not for local files) which has a performance cost.

Also removes the "gantry" boolean because I think we can use `data_root:/` if we want that

### CODE REVIEW

**Comment by jder** [comment]
I should probably add some tests for the Location types but thought I'd get your feedback first!

File: src/ocean_emulators/config.py [hunk 1]
```diff
@@ -185,17 +200,13 @@ def nets_dir(self) -> Path:
...
-            return Path("/")
-        else:
-            if self.cluster_data_dir is None:
-                raise ValueError(
-                    "cluster_data_dir must be set, try"
-                    " --experiment.cluster_data_dir=path/to/data"
-                )
-            else:
-                return Path(self.cluster_data_dir)
+    def resolved_data_root(self) -> ResolvedLocation:
```
Comment (alxmrs) [inline src/ocean_emulators/config.py:203]: I like this runtime validation and type enforcement!

File: src/ocean_emulators/utils/data.py [hunk 2]
```diff
@@ -230,24 +228,16 @@ def from_dataset(cls, dataset: xr.Dataset, *, name: str = "dataset") -> Self:
...
-            chunks,
-        )
-        stds = _open_dataset(
-            cfg.data.data_stds_path,
-            root,
-            chunks,
-        )
+        data = root.resolve(cfg.data.data_location).open(chunks)
+        means = root.resolve(cfg.data.data_means_location).open(chunks)
+        stds = root.resolve(cfg.data.data_stds_location).open(chunks)
```
Comment (alxmrs) [inline src/ocean_emulators/utils/data.py:235]: This is slick!

File: src/ocean_emulators/utils/data.py [hunk 3]
```diff
@@ -230,24 +228,16 @@ def from_dataset(cls, dataset: xr.Dataset, *, name: str = "dataset") -> Self:
...
 
         dask = "with_dask" if use_dask else "without_dask"
 
         return cls(
-            name=f"{cfg.experiment.name}-{cfg.experiment.data_dir.name}-{dask}",
+            name=f"{cfg.experiment.name}-{root}-{dask}",
```
Comment (alxmrs) [inline src/ocean_emulators/utils/data.py]: I love how succinct this is now.

File: src/ocean_emulators/utils/location.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+class ResolvedLocation(ABC):
+    """A location which is ready to be opened or resolved against."""
+
+    @abstractmethod
+    def open(self, chunks: dict[str, int] | None = None) -> xr.Dataset:
+        pass
+
+    @abstractmethod
+    def resolve(self, location: "Location") -> "ResolvedLocation":
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:42]: 🐑  Does this work here?
```suggestion
    def resolve(self, location: "Location") -> Self:
```

File: src/ocean_emulators/utils/location.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        pass
+
+    def __truediv__(self, other: "Location") -> "ResolvedLocation":
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:45]: 🐑 
```suggestion
    def __truediv__(self, other: "Location") -> Self:
```

File: src/ocean_emulators/utils/location.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    def open(self, chunks: dict[str, int] | None = None) -> xr.Dataset:
+        # TODO(jder): could consider passing credentials here
+        # rather than relying on the environment
+
+        return xr.open_dataset(
+            self.url(),
+            backend_kwargs={"storage_options": {"endpoint_url": self.endpoint_url}},
+            engine="zarr",
+            chunks=chunks,
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:74]: I believe `chunks=None` and `chunks={}` are different behaviors and I think we want the latter -- unless: do you mean to defer opening the Zarr data?

File: src/ocean_emulators/utils/location.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    endpoint_url: str | None = None
+    bucket: str
+    path: str
+
+    def open(self, chunks: dict[str, int] | None = None) -> xr.Dataset:
+        # TODO(jder): could consider passing credentials here
+        # rather than relying on the environment
+
+        return xr.open_dataset(
+            self.url(),
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:71]: I love this interface!

File: src/ocean_emulators/utils/location.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            backend_kwargs={"storage_options": {"endpoint_url": self.endpoint_url}},
+            engine="zarr",
+            chunks=chunks,
+        )
+
+    def url(self) -> str:
+        path = quote(self.path.lstrip("/"))
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:78]: TIL about the `quote` function.

File: src/ocean_emulators/utils/location.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        pass
+
+    def __truediv__(self, other: "Location") -> "ResolvedLocation":
+        return self.resolve(other)
+
+
+class S3Location(ResolvedLocation, BaseModel):
+    """An S3 bucket, assuming credentials in your environment.
+
+    For example:
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:52]: Thanks for including the yaml config example, that's useful.

File: src/ocean_emulators/utils/location.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      path: /path/to/data
+    ```
+    """
+
+    type: Literal["local"] = "local"
+    path: Path
+
+    def open(self, chunks: dict[str, int] | None = None) -> xr.Dataset:
+        engine = "netcdf4" if self.path.suffix == ".nc" else "zarr"
+        return xr.open_dataset(self.path, engine=engine, chunks=chunks)
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:111]: See the other `chunks` note above. Maybe at the call sight we pass `chunks={}`? I need to double check.

File: src/ocean_emulators/utils/location.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+      type: local
+      path: /path/to/data
+    ```
+    """
+
+    type: Literal["local"] = "local"
+    path: Path
+
+    def open(self, chunks: dict[str, int] | None = None) -> xr.Dataset:
+        engine = "netcdf4" if self.path.suffix == ".nc" else "zarr"
```
Comment (alxmrs) [inline src/ocean_emulators/utils/location.py:110]: I like how this is adapted for how we use the system. I think it's still poignant in that NetCDF is a local fs format anyway.

**Comment by alxmrs** [comment]
Your approach is quite type-safe. I like how you create abstractions from the same concept based around how they're used. I think you do a good job at building affordances around how we care they're used rather than the varying components for how data locations could be specified. Good job and I look forward to seeing this PR fully-baked.

**Comment by alxmrs** [comment]
Very rusty, well done.

File: src/ocean_emulators/utils/location.py (hunk 4 above)
Comment (jder) [reply src/ocean_emulators/utils/location.py:42]: Yeah, I almost did that but we can't promise that resolve returns the class of `self` -- if you have `self: LocalLocation` and `location: S3Location` then `self.resolve(location)` returns a S3Location, not a LocalLocation (in this case, Self would resolve to a LocalLocation). All we can promise is that it's *some* `ResolvedLocation`, not necessarily the same type as the instance you're calling it on.

File: src/ocean_emulators/utils/location.py (hunk 6 above)
Comment (jder) [reply src/ocean_emulators/utils/location.py:74]: Yeah, at the callsite we pass `{}` when we want dask. This is the same default as xarray.open_dataset, I think?

File: src/ocean_emulators/utils/location.py (hunk 4 above)
Comment (jder) [reply src/ocean_emulators/utils/location.py:42]: (Sorry, I ninja-edited my comment when I went to actually go do this and realized we can't)

**Comment by jder** [comment]
@alxmrs I added some tests and rebased on the latest main, do you want to take another look?

File: src/ocean_emulators/utils/location.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/location.py:74]: Yeah, I think you're right about the default -- further, I'm happy this is at the callsite. Thanks for confirming.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -349,7 +349,7 @@ uv run pytest --benchmark-only --benchmark-cprofile="tottime_per" --benchmark-cp
 (Please consult [these docs](https://pytest-benchmark.readthedocs.io/en/latest/usage.html#:~:text=%2D%2Dbenchmark%2Dcprofile%3DCOLUMN)
 to see all available values for the `--benchmark-cprofile` flag.)
 
-This will generate a `.prof` file located in the `` directory (by default). You can visualize this trace with `snakeviz`
+This will generate a `.prof` file located in the `.` directory (by default). You can visualize this trace with `snakeviz`
```
Comment (alxmrs) [inline CONTRIBUTING.md:352]: nice, thanks

File: configs/data/cm4.yaml [hunk 13]
```diff
@@ -1,8 +1,8 @@
 # yaml-language-server: $schema=../schemas/DataConfig.json
 
-data_path: cm4_piControl_ocean_200yr_full_chunked.zarr
-data_means_path: 2024-11-12-cm4-piControl-ocean-200yr-dataset-stats/centering.nc
-data_stds_path: 2024-11-12-cm4-piControl-ocean-200yr-dataset-stats/scaling-full-field.nc
+data_location: cm4_piControl_ocean_200yr_full_chunked.zarr
+data_means_location: 2024-11-12-cm4-piControl-ocean-200yr-dataset-stats/centering.nc
+data_stds_location: 2024-11-12-cm4-piControl-ocean-200yr-dataset-stats/scaling-full-field.nc
```
Comment (alxmrs) [inline configs/data/cm4.yaml:5]: As a random aside (for a future PR): I think storing the data + means + std in an Xarray DataTree would be a good way to go. See: https://github.com/Open-Athena/hermes/blob/main/tests/test_samudra.py

File: configs/data/om4_halfdeg_remote.yaml [hunk 14]
```diff
@@ -1,10 +1,23 @@
 # yaml-language-server: $schema=../schemas/DataConfig.json
 
-# TODO(jder): the URLs below assume you have FSSPEC_S3_ENDPOINT_URL=https://nyu1.osn.mghpcc.org
-# as well as the appropriate AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY (or other S3 credentials) in your environment.
-# Once we have a clear story for launching this task remotely we should remove this implicit requirement.
+# TODO(jder): we assume you have appropriate AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY (or other S3 credentials) in your environment.
```
Comment (alxmrs) [inline configs/data/om4_halfdeg_remote.yaml]: I've been putting these throughout the codebase, WDYT?

```suggestion
# NB(jder): we assume you have appropriate AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY (or other S3 credentials) in your environment.
```

Comment (alxmrs) [reply configs/data/om4_halfdeg_remote.yaml]: Taking a leaf from your book ;)

File: configs/data/om4_halfdeg_remote.yaml [hunk 15]
```diff
@@ -1,10 +1,23 @@
...
+
+data_location:
+  type: s3
+  bucket: emulators
+  path: sd5313/OM4_highres/om4_halfdeg.zarr
+  endpoint_url: https://nyu1.osn.mghpcc.org
```
Comment (alxmrs) [inline configs/data/om4_halfdeg_remote.yaml:10]: Really nice!

File: configs/train_default.test.yaml [hunk 16]
```diff
@@ -42,11 +42,11 @@ experiment:
   network: Samudra
   prognostic_vars_key: thermo_dynamic_5
   boundary_vars_key: tau_hfds
-  cluster_data_dir: .data_cache/
+  data_root: .data_cache/
```
Comment (alxmrs) [inline configs/train_default.test.yaml:45]: `data_root` is so much clearer + accurate.

File: tests/test_location.py [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+
+    def test_valid_absolute_path(self):
+        """Test that absolute paths are accepted."""
+        loc = UnresolvedLocation(path="/absolute/path/data.zarr")
+        assert loc.path == "/absolute/path/data.zarr"
+
+    def test_invalid_absolute_url(self):
+        """Test that absolute URLs are rejected."""
+        with pytest.raises(ValidationError, match="Absolute urls are not supported"):
+            UnresolvedLocation(path="s3://bucket/path")
```
Comment (alxmrs) [inline tests/test_location.py:32]: Nice to see a test for this, thanks.

File: tests/test_location.py [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+        assert resolved.path == Path("/base/path/subdir/file.zarr")
+
+    def test_resolve_resolved_location(self):
+        """Test resolving a ResolvedLocation returns it unchanged."""
+        base = LocalLocation(path=Path("/base/path"))
+        other = LocalLocation(path=Path("/other/path"))
+
+        resolved = base.resolve(other)
+
+        assert resolved is other
```
Comment (alxmrs) [inline tests/test_location.py:81]: Nice `__eq__` test

File: tests/test_location.py [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+            ds.to_netcdf(nc_path)
+
+            # Test opening
+            loc = LocalLocation(path=nc_path)
+            opened_ds = loc.open()
+
+            assert isinstance(opened_ds, xr.Dataset)
+            assert "temperature" in opened_ds.data_vars
+
+    def test_open_with_chunks(self):
```
Comment (alxmrs) [inline tests/test_location.py:123]: 🐑  optional nit: might be nice to add a test to check chunks with values, e.g. `chunks=dict(x=2)`, etc. 

Though, I do think that such a feature works.

Comment (alxmrs) [reply tests/test_location.py:123]: You can assert that `ds.chunks` equates to what you expect.

File: tests/test_location.py [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+
+    def test_url_generation(self):
+        """Test URL generation for S3Location."""
+        loc = S3Location(bucket="test-bucket", path="data/test.zarr")
+        assert loc.url() == "s3://test-bucket/data/test.zarr"
+
+    def test_url_generation_with_special_chars(self):
+        """Test URL generation with special characters."""
+        loc = S3Location(bucket="test-bucket", path="data/test file.zarr")
+        assert loc.url() == "s3://test-bucket/data/test%20file.zarr"
```
Comment (alxmrs) [inline tests/test_location.py:161]: Nice! Great test.

File: tests/test_location.py [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+
+            # Test opening with chunks
+            loc = LocalLocation(path=zarr_path)
+            opened_ds = loc.open(chunks={})
+
+            assert isinstance(opened_ds, xr.Dataset)
+            assert "temperature" in opened_ds.data_vars
+
+
+class TestS3Location:
```
Comment (alxmrs) [inline tests/test_location.py:141]: 🐑  Including an explicit test of the `url()`, specifically of `self.path.lstrip("/")` might be nice. Even just prepending a `/` to one of the tests blow should do it -- and, maybe this exists but I've missed it.

File: tests/test_location.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+
+        resolved = base.resolve(unresolved)
+        assert isinstance(resolved, S3Location)
+
+        assert resolved.endpoint_url == "https://s3.example.com"
+
+    def test_resolve_resolved_location(self):
+        """Test resolving a ResolvedLocation returns it unchanged."""
+        base = S3Location(bucket="test-bucket", path="base/path")
+        other = LocalLocation(path=Path("/other/path"))
```
Comment (alxmrs) [inline tests/test_location.py:197]: I think this may exercise the `url()` test that I mentioned above.

File: tests/test_location.py [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+
+        unresolved = UnresolvedLocation(path="data/test.zarr")
+        model = TestModel(location=unresolved)
+        assert model.location is unresolved
+
+
+class TestLocationIntegration:
+    """Integration tests for the location system."""
+
+    def test_mixed_location_types(self):
```
Comment (alxmrs) [inline tests/test_location.py:244]: It's cool how we can mix these!

File: tests/test_location.py [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+        assert s3_resolved.path == "base/subdir/file.zarr"
+
+        # Test local resolution
+        local_resolved = base_local.resolve(unresolved)
+        assert isinstance(local_resolved, LocalLocation)
+        assert local_resolved.path == Path("/local/base/subdir/file.zarr")
+
+    def test_path_joining_edge_cases(self):
+        """Test edge cases in path joining."""
+        # Test with trailing slash on base
```
Comment (alxmrs) [inline tests/test_location.py:262]: Ah, it looks like you already got address the thing I called out! I should have kept scrolling.

**Review by alxmrs — APPROVED** [review]
Great work, LGTM

File: configs/data/om4_halfdeg_remote.yaml (hunk 14 above)
Comment (jder) [reply configs/data/om4_halfdeg_remote.yaml]: How is "NB" different from just a normal comment? I meant "TODO" insofar as we should at some point resolve that assumption.

Comment (alxmrs) [reply configs/data/om4_halfdeg_remote.yaml]: Ah, I see how it is an action item to be addressed, I didn't understand it that way at first. 

I think I distinguish NB from a normal comment in that it's tagged with a username. But, that's not a long standing convention or anything, just a pattern I've come to like.

Comment (jder) [reply configs/data/om4_halfdeg_remote.yaml]: Gotcha, I updated the comment to make it clear what there is to be done.
